### PR TITLE
CacheDetailActivity: improve color detection

### LINKED ISSUE #7153: Don't process color tags in cache description
##### Detailed steps causing the problem:
- Open some cache with color formatting (i.e. MS20692)
- Check description
- Make sure colors are being rendered by c:geo which results into mostly non-readable text (blue text on black background

Please either disable colors at all (so far I haven't seen any cache where colors were important) or at least add option "don't render colors"

### PR DESCRIPTION
**Pull request by okainov**
Use W3 recommendations for determining contrast to decide whether to use neutral gray color.

Reference
https://www.w3.org/TR/2008/REC-WCAG20-20081211/#visual-audio-contrast-contrast

Fixes #7153

### CODE REVIEW

**Comment by Lineflyer** [comment]
ok to test

File: tests/src-android/cgeo/geocaching/utils/ColorUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package cgeo.geocaching.utils;
+
+import android.graphics.Color;
+
+import junit.framework.TestCase;
+
+import static org.assertj.core.api.Java6Assertions.assertThat;
```
[the remark this answers is not in the document]
Comment (okainov) [reply tests/src-android/cgeo/geocaching/utils/ColorUtilsTest.java]: I just used Androd Studio "Reformat code \ Optimize Import"... Changed the settings now, done

File: main/src/cgeo/geocaching/utils/ColorUtils.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    /**
+     * Calculates contrast between two {@link Color} int according to W3
+     * <p>
+     * See https://www.w3.org/TR/2008/REC-WCAG20-20081211/#contrast-ratiodef
+     *
+     * @param rgb1 first {@link Color} int
+     * @param rgb2 second {@link Color} int
+     * @return contrast in range from 1 to 21
+     */
+    public static double getContrast(final int rgb1, final int rgb2) {
```
[the remark this answers is not in the document]
Comment (okainov) [reply main/src/cgeo/geocaching/utils/ColorUtils.java]: done

File: main/src/cgeo/geocaching/CacheDetailActivity.java [hunk 3]
```diff
@@ -1,5 +1,82 @@
...
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.regex.Pattern;
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
```
[the remark this answers is not in the document]
Comment (okainov) [reply main/src/cgeo/geocaching/CacheDetailActivity.java]: I hope, done

File: main/src/cgeo/geocaching/CacheDetailActivity.java [hunk 4]
```diff
@@ -175,21 +177,14 @@
...
-    private static final Pattern[] LIGHT_COLOR_PATTERNS = {
-            // html attributes
-            Pattern.compile("((?<!bg)color)=\"#" + "([F][6-9A-F]){3}" + "\"", Pattern.CASE_INSENSITIVE),
-            Pattern.compile("((?<!bg)color)=\"" + "white" + "\"", Pattern.CASE_INSENSITIVE),
-            // styles
-            Pattern.compile("((?<!background-)color):#" + "([F][6-9A-F]){3}" + COLOR_END, Pattern.CASE_INSENSITIVE), Pattern.compile("((?<!background-)color):" + "white" + COLOR_END, Pattern.CASE_INSENSITIVE) };
+
+    private static final Pattern[] COLOR_PATTERNS = {
+            // <span style="color: #0000ff;">
+            Pattern.compile("(?<!background-)color: ?\"?(#[0-9a-f]{6}|\\w+)", Pattern.CASE_INSENSITIVE),
```
[the remark this answers is not in the document]
Comment (okainov) [reply main/src/cgeo/geocaching/CacheDetailActivity.java]: Quote sign is missing in the suggestion, but for whitespace yes, replaced.

File: main/src/cgeo/geocaching/CacheDetailActivity.java [hunk 5]
```diff
@@ -175,21 +177,14 @@
...
+            // <font color="red">
+            Pattern.compile("(?<!bg)color= ?\"?(#[0-9a-f]{6}|\\w+)", Pattern.CASE_INSENSITIVE)};
```
[the remark this answers is not in the document]
Comment (okainov) [reply main/src/cgeo/geocaching/CacheDetailActivity.java]: done

File: main/src/cgeo/geocaching/CacheDetailActivity.java [hunk 6]
```diff
@@ -1782,29 +1777,28 @@ private void addWarning(final UnknownTagsHandler unknownTagsHandler, final Spann
...
-            for (final Pattern pattern : DARK_COLOR_PATTERNS) {
-                final MatcherWrapper matcher = new MatcherWrapper(pattern, descriptionString);
-                if (matcher.find()) {
-                    descriptionView.setBackgroundResource(color.darker_gray);
-                    return;
+        for (final Pattern colorPattern : COLOR_PATTERNS) {
+            final MatcherWrapper matcher = new MatcherWrapper(colorPattern, descriptionString);
+            while (matcher.find()) {
+                try {
+                    final int textColor = Color.parseColor(matcher.group(0));
```
[the remark this answers is not in the document]
Comment (okainov) [reply main/src/cgeo/geocaching/CacheDetailActivity.java]: done

File: main/src/cgeo/geocaching/CacheDetailActivity.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (UniQP) [reply main/src/cgeo/geocaching/CacheDetailActivity.java]: The `org.apache.commons` imports are still moved to a undesired position.

File: main/src/cgeo/geocaching/CacheDetailActivity.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (UniQP) [reply main/src/cgeo/geocaching/CacheDetailActivity.java]: Is a quote sign allowed at this place?

File: main/src/cgeo/geocaching/CacheDetailActivity.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (okainov) [reply main/src/cgeo/geocaching/CacheDetailActivity.java]: done

File: main/src/cgeo/geocaching/CacheDetailActivity.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (okainov) [reply main/src/cgeo/geocaching/CacheDetailActivity.java]: Probably missed this point, thanks.

**Comment by okainov** [comment]
By manually going thru my lists, found a few examples:

https://coord.info/GC1Y4B8 - "usually have an oval" is written in black
https://www.geocaching.com/geocache/GCM03P_united-colors?guid=89cec7d3-7c3d-4096-afe6-45eb4a0615d7 this looks ugly with explicit black font.

However I'm not a big expert in COM caches, so maybe @SammysHP can share some of hundreds caches with colors he had seen.

File: main/src/cgeo/geocaching/CacheDetailActivity.java [hunk 7]
```diff
@@ -1734,7 +1729,7 @@ private void loadDescription(final String descriptionString, final IndexOutOfBou
 
                     fixRelativeLinks(descriptionView);
                     descriptionView.setMovementMethod(AnchorAwareLinkMovementMethod.getInstance());
-                    fixTextColor(descriptionString, descriptionView);
+                    fixTextColor(Html.toHtml(description), descriptionView);
```
Comment (UniQP) [inline main/src/cgeo/geocaching/CacheDetailActivity.java]: Can you add a comment that explains why not to use `descriptionString` (similar to the explanation in the commit message).

Comment (okainov) [reply main/src/cgeo/geocaching/CacheDetailActivity.java]: Just found out that there is already proper span parsing in fixRelativeUrl method, so I thew away all HTML parsing with regexps and replaced it to neat span parsing. Sorry for several iterations

**Comment by Lineflyer** [comment]
@UniQP If you are fine with the code itself, feel free to merge to test it in nightlies. I will scan certain listings and see if the behaviour improved.

**Comment by okainov** [comment]
@UniQP do you want me to squash commits at some point to have more neat history? I saw sometimes you merge it without squashing (usually GitHub allows you to merge PR and squash automatically), so it's not really clear to have 15 commits like "fix code style"...

**Comment by UniQP** [comment]
@okainov I tested the code with GCM03P and it seems that setting the background color doesn't work.

> @UniQP do you want me to squash commits at some point to have more neat history? I saw sometimes you merge it without squashing […]

I will merge without squashing only if the commit history is clean (e.g. contains no fixup commits). You can decide yourself if you want to create a clean commit history in the end or if squashing all changes into a single commit is the right thing to do.

**Comment by okainov** [comment]
>I tested the code with GCM03P and it seems that setting the background color doesn't work.

Good catch, fixed it now, postprocessing was done after setting the text.

I squashed and rebased all the commits since they all were related to this PR.

**Comment by UniQP** [comment]
I continued testing and noticed that CSS colors aren't working anymore. The reason is that `Spannable` does not support `<style>` tags and CSS colors.

**Comment by okainov** [comment]
@UniQP could you please provide cache where you observe this?

**Comment by UniQP** [comment]
Strange: I tested GC54REW where I didn't see an effect. On the other hand, it works for GC2PAEP.

**Comment by okainov** [comment]
>Strange: I tested GC54REW where I didn't see an effect. On the other hand, it works for GC2PAEP.

GC54REW has `<div style="background-color:#8597BB;color:#000000` in its descriptionString.
GC2PAEP has usual spans `<span style="color:#000000;">` in its descriptionString.

Html.toHtml(Html.fromHtml(string)) removes all the divs because it's not supported. It doesn't depend on this PR.

Original implementation behavior:
GC54REW - triggers gray background.  However it's useless, because, as noticed before, color in divs is not supported by Html.fromHtml. So even if one explicitly disables `fixColorText`, content will be perfectly readable.
GC2PAEP  - no fix, only black text, not readable. Reason why it wan't triggering on "span style" - lack of space in regexp betweeen ":" and "#00"

This PR implementation behavior:
GC54REW - no fix, since there is no color problems.
GC2PAEP - triggers gray background for part of description which is inside  `<span style="color:#000000;">`

**Comment by UniQP** [comment]
Thank you for the investigation. I agree that it's a good thing to support exactly the HTML tags that are supported by the TextView.

**Comment by okainov** [comment]
@rsudev @UniQP @SammysHP @Lineflyer  Just a friendly reminder, any updates?

**Pull request merged**